### PR TITLE
[tree] Propagate TTreeIndex info through GetFriendInfo/MakeFriends 

### PR DESCRIPTION
**Pull request by eguiraud**
This in turn fixes a problem with TTreeProcessorMT and multi-thread
RDataFrame "forgetting" about the TTreeIndexes associated with
input friend trees.

It fixes https://github.com/root-project/root/issues/12260,
"[DF] Bogus data read from indexed friend trees in multi-thread runs".

A test is added for this case as well.

### CODE REVIEW

File: tree/tree/src/InternalTreeUtils.cxx [hunk 1]
```diff
@@ -200,6 +202,12 @@ ROOT::TreeUtils::RFriendInfo GetFriendInfo(const TTree &tree, bool retrieveEntri
       const auto *alias_c = tree.GetFriendAlias(frTree);
       const std::string alias = alias_c != nullptr ? alias_c : "";
 
+      auto *treeIndex = frTree->GetTreeIndex();
+      if (treeIndex)
```
Comment (vepadulano) [inline tree/tree/src/InternalTreeUtils.cxx:206]: Maybe it's  also worth checking that `!treeIndex->IsZombie()` since apparently that's done in [BuildIndex](https://github.com/root-project/root/blob/cca7c59a0878ebd7942c6a16516dd34bbb23d912/tree/tree/src/TTree.cxx#L2638) as well

File: tree/tree/src/InternalTreeUtils.cxx [hunk 2]
```diff
@@ -382,6 +391,11 @@ std::vector<std::unique_ptr<TChain>> MakeFriends(const ROOT::TreeUtils::RFriendI
             frChain->Add((thisFriendFileNames[j] + "?#" + thisFriendChainSubNames[j]).c_str(), thisFriendEntries[j]);
          }
       }
+
+      const auto &indexInfo = finfo.fTreeIndexInfos[i];
```
Comment (vepadulano) [inline tree/tree/src/InternalTreeUtils.cxx:395]: To make sure that `finfo.fTreeIndexInfos[i]` always exists, we should probably also add an empty `TreeIndexInfo` in all the `RFriendInfo::AddFriend` overloads

Comment (vepadulano) [reply tree/tree/src/InternalTreeUtils.cxx:395]: This is probably not relevant for `TTreeProcessorMT` but it would be relevant for `RLoopManager(RDatasetSpec &&spec)`. See https://github.com/root-project/root/pull/12262 , tests were failing due to the missing addition of the friend entries in the `RFriendInfo::AddFriend` overloads. This would probably mean that as soon as that PR is merged also this one would start showing the same failures

**Review by vepadulano — CHANGES_REQUESTED** [review]
Nice catch! See a few minor comments

File: tree/tree/src/InternalTreeUtils.cxx (hunk 1 above)
Comment (eguiraud) [reply tree/tree/src/InternalTreeUtils.cxx:206]: ah, I didn't see that, thanks! but I think precisely because `BuildIndex` checks for `isZombie()` on the newly-constructed `TTreeIndex` and does not attach it to the `TTree` if it is then there is no code path where a TTree has a zombie TTreeIndex attached (as it should be).

File: tree/tree/src/InternalTreeUtils.cxx [hunk 3]
```diff
@@ -268,7 +276,8 @@ ROOT::TreeUtils::RFriendInfo GetFriendInfo(const TTree &tree, bool retrieveEntri
    }
 
    return ROOT::TreeUtils::RFriendInfo{std::move(friendNames), std::move(friendFileNames),
-                                       std::move(friendChainSubNames), std::move(nEntriesPerTreePerFriend)};
+                                       std::move(friendChainSubNames), std::move(nEntriesPerTreePerFriend),
+                                       std::move(treeIndexes)};
```
Comment (pcanal) [inline tree/tree/src/InternalTreeUtils.cxx:280]: `treeIndexes` is a pointer.  What is the point being made here by using `std::move`?

File: tree/tree/src/InternalTreeUtils.cxx [hunk 4]
```diff
@@ -382,6 +391,11 @@ std::vector<std::unique_ptr<TChain>> MakeFriends(const ROOT::TreeUtils::RFriendI
             frChain->Add((thisFriendFileNames[j] + "?#" + thisFriendChainSubNames[j]).c_str(), thisFriendEntries[j]);
          }
       }
+
+      const auto &indexInfo = finfo.fTreeIndexInfos[i];
+      if (indexInfo.fHasIndexInfo)
+         frChain->BuildIndex(indexInfo.fMajorName.c_str(), indexInfo.fMinorName.c_str());
```
Comment (pcanal) [inline tree/tree/src/InternalTreeUtils.cxx:397]: If `frChain` is indeed a chain, this implies that every single tree (and file) of the chain will be traverse 'right here and now', is that the intent?

File: tree/tree/src/InternalTreeUtils.cxx (hunk 3 above)
Comment (eguiraud) [reply tree/tree/src/InternalTreeUtils.cxx:280]: `treeIndexes` is a `std::vector<TreeIndexInfos>`

Comment (pcanal) [reply tree/tree/src/InternalTreeUtils.cxx:280]: I see. I got confused by `treeIndex` vs `treeIndexes` while scanning the diff.

File: tree/tree/src/InternalTreeUtils.cxx (hunk 4 above)
Comment (eguiraud) [reply tree/tree/src/InternalTreeUtils.cxx:397]: uhm I guess that instead of storing the majorname and minorname and call `BuildIndex` again in every task I could store the TTreeIndex itself and attach a copy of it to every chain in every task?

File: tree/tree/src/InternalTreeUtils.cxx (hunk 2 above)
Comment (eguiraud) [reply tree/tree/src/InternalTreeUtils.cxx:395]: should be done

File: tree/tree/src/InternalTreeUtils.cxx (hunk 4 above)
Comment (eguiraud) [reply tree/tree/src/InternalTreeUtils.cxx:397]: I couldn't manage to make it work, partly because `TChainIndex::SetTree` does not actually set `fTree`, partly because I'm not sure what happens if I copy a TTreeIndex out of a TTree and then assign it to a TChain.

Comment (pcanal) [reply tree/tree/src/InternalTreeUtils.cxx:397]: > if I copy a TTreeIndex out of a TTree and then assign it to a TChain.

That does not seem to make sense, wouldn't the entry be off or not enough?

>  again in every task I could store the TTreeIndex 

For a tree index, you could, for example `TTree uses`:
```
               TVirtualIndex *index = (TVirtualIndex*) oldtree->GetTree()->GetTreeIndex()->Clone();
               index->SetTree(newtree);
               newtree->SetTreeIndex(index);
```
For a `TChainIndex` that is more dubious (beside that `TChainIndex::SetTree` would have be actually implemented) because apriori the index is very large (because it ought to an an entry for every single entry of all the trees in the chian, isn't it?) and duplicating it sounds ... less than optimal.

Comment (pcanal) [reply tree/tree/src/InternalTreeUtils.cxx:397]: `TChainIndex`, looks like it contains 'just' the range of the trees and the index for those trees that do not already have an index (so it could still be pretty large).  Ideally those would be created only once (in a distributed fashion if possible) and shared.  (It will requires updates to `TChainIndex` to make it so).

Comment (eguiraud) [reply tree/tree/src/InternalTreeUtils.cxx:397]: mmh that sounds way too time consuming...

Comment (eguiraud) [reply tree/tree/src/InternalTreeUtils.cxx]: `TChainIndex::SetTree` was changed to actually set the tree and a new version of the fix that simply clones the original TTreeIndex/TChainIndex instead of re-building every time has been pushed. Things seem to work.

I think we are missing a test for the corner case in which we have a TTree in input rather than a TChain, I'll add it soon.

Comment (eguiraud) [reply tree/tree/src/InternalTreeUtils.cxx]: ...tests added

File: tree/tree/inc/ROOT/RFriendInfo.hxx [hunk 5]
```diff
@@ -58,13 +63,19 @@ struct RFriendInfo {
     */
    std::vector<std::vector<std::int64_t>> fNEntriesPerTreePerFriend;
 
-   void AddFriend(const std::string &treeName, const std::string &fileNameGlob, const std::string &alias = "");
+   /**
+      Information on the friend's TTreeIndexes.
+   */
+   std::vector<TVirtualIndex *> fTreeIndexInfos;
```
Comment (vepadulano) [inline tree/tree/inc/ROOT/RFriendInfo.hxx:69]: Pass-by comment. It's scary IMO to see a vector of raw pointers in a new public interface. I see the point in having it to avoid copies, but I wonder if this is the only way. If it is, I would spend some wording in the docs to discuss the lifetime and ownership of the `TVirtualIndex *` inside the vector.

Comment (eguiraud) [reply tree/tree/inc/ROOT/RFriendInfo.hxx:69]: Mmmh thank you for pointing this out. The raw pointers are non-owning and point to the tree indexes in the original friends (which in the RDF/TTreeProcessorMT case must have lifetimes at least as long as the event loop, so this saves us a copy for the local case). For the distributed case, if we stream the RFriendInfo to the tasks, I think we'll leak the tree indexes.

Comment (vepadulano) [reply tree/tree/inc/ROOT/RFriendInfo.hxx:69]: > For the distributed case, if we stream the RFriendInfo to the tasks, I think we'll leak the tree indexes.

Yes I was also thinking about that. Currently we don't support TTreeIndex in distributed mode, but this would make it simply impossible to support I believe.

Comment (eguiraud) [reply tree/tree/inc/ROOT/RFriendInfo.hxx]: switched to a `vector<unique_ptr<TVirtualIndex>>` although it makes the patch a bit more comlpicated

File: tree/treeplayer/src/TChainIndex.cxx [hunk 6]
```diff
@@ -182,9 +182,11 @@ void TChainIndex::DeleteIndices()
 
 TChainIndex::~TChainIndex()
 {
-   DeleteIndices();
-   if (fTree && fTree->GetTreeIndex() == this)
-      fTree->SetTreeIndex(0);
+   if (fTree) {
+      DeleteIndices();
+      if (fTree->GetTreeIndex() == this)
+         fTree->SetTreeIndex(0);
```
Comment (vepadulano) [inline tree/treeplayer/src/TChainIndex.cxx:188]: Take the chance to change this to `nullptr` perhaps?

File: tree/tree/inc/ROOT/RFriendInfo.hxx [hunk 7]
```diff
@@ -58,13 +63,30 @@ struct RFriendInfo {
...
 
-   void AddFriend(const std::string &treeName, const std::string &fileNameGlob, const std::string &alias = "");
+   /**
+      Information on the friend's TTreeIndexes.
+   */
+   std::vector<std::unique_ptr<TVirtualIndex>> fTreeIndexInfos;
+
+   RFriendInfo() = default;
+   RFriendInfo(const RFriendInfo &);
+   RFriendInfo &operator=(const RFriendInfo &);
```
Comment (vepadulano) [inline tree/tree/inc/ROOT/RFriendInfo.hxx:73]: This has become a heavy struct, I wonder if we can avoid enabling copy semantics. In what case do we need to copy the friend info around once created?

File: tree/tree/inc/ROOT/RFriendInfo.hxx [hunk 8]
```diff
@@ -58,13 +63,30 @@ struct RFriendInfo {
...
+   RFriendInfo(RFriendInfo &&) = default;
+   RFriendInfo &operator=(RFriendInfo &&) = default;
+   RFriendInfo(std::vector<std::pair<std::string, std::string>> friendNames,
+               std::vector<std::vector<std::string>> friendFileNames,
+               std::vector<std::vector<std::string>> friendChainSubNames,
+               std::vector<std::vector<std::int64_t>> nEntriesPerTreePerFriend,
+               std::vector<std::unique_ptr<TVirtualIndex>> treeIndexInfos);
```
Comment (vepadulano) [inline tree/tree/inc/ROOT/RFriendInfo.hxx:80]: I guess this constructor is needed to allow for the other constructors/assignment operators right? In principle, if copy semantics are removed, all the constructors could be removed too IIUC

File: tree/tree/src/InternalTreeUtils.cxx [hunk 9]
```diff
@@ -200,6 +202,9 @@ ROOT::TreeUtils::RFriendInfo GetFriendInfo(const TTree &tree, bool retrieveEntri
       const auto *alias_c = tree.GetFriendAlias(frTree);
       const std::string alias = alias_c != nullptr ? alias_c : "";
 
+      auto *treeIndex = frTree->GetTreeIndex();
+      treeIndexes.emplace_back(static_cast<TVirtualIndex *>(treeIndex ? treeIndex->Clone() : nullptr));
```
Comment (vepadulano) [inline tree/tree/src/InternalTreeUtils.cxx:206]: Right so now we `Clone` irrespectively of being in local/distributed mode. It's an extra operation but allows for the distributed case

File: tree/tree/src/InternalTreeUtils.cxx [hunk 10]
```diff
@@ -382,6 +388,14 @@ std::vector<std::unique_ptr<TChain>> MakeFriends(const ROOT::TreeUtils::RFriendI
             frChain->Add((thisFriendFileNames[j] + "?#" + thisFriendChainSubNames[j]).c_str(), thisFriendEntries[j]);
          }
       }
+
+      auto &treeIndex = finfo.fTreeIndexInfos[i];
+      if (treeIndex) {
+         auto *copyOfIndex = static_cast<TVirtualIndex *>(treeIndex->Clone());
+         copyOfIndex->SetTree(frChain.get());
+         frChain->SetTreeIndex(copyOfIndex);
+      }
```
Comment (vepadulano) [inline tree/tree/src/InternalTreeUtils.cxx:397]: IIUC now the `fTreeIndexInfos` owns its `TVirtualIndex`es. Why do we need to `Clone` again here?

File: tree/tree/inc/ROOT/RFriendInfo.hxx (hunk 7 above)
Comment (eguiraud) [reply tree/tree/inc/ROOT/RFriendInfo.hxx:73]: Users of the struct should take care of moving instead of copying when possible (as it happens commonly with user-defined types), but I don't think it's the responsibility of the struct to completely disallow copies. Also it would be a backward-incompatible change as RFriendInfo is in the public namespace -- I wouldn't mix it with this PR which is only meant to fix #12260 .

File: tree/tree/inc/ROOT/RFriendInfo.hxx (hunk 8 above)
Comment (eguiraud) [reply tree/tree/inc/ROOT/RFriendInfo.hxx:80]: > this constructor is needed to allow for the other constructors/assignment operators right?

No, this constructor is needed by `GetFriendInfo`. The other constructors or assignment operators can live without this.

File: tree/tree/src/InternalTreeUtils.cxx (hunk 10 above)
Comment (eguiraud) [reply tree/tree/src/InternalTreeUtils.cxx:397]: Each friend tree/chain in each task needs its own copy of the thing.

File: tree/tree/inc/ROOT/RFriendInfo.hxx (hunk 8 above)
Comment (vepadulano) [reply tree/tree/inc/ROOT/RFriendInfo.hxx:80]: It's just that previously there were no constructors so I was wondering why now we need one.

File: tree/tree/inc/ROOT/RFriendInfo.hxx (hunk 7 above)
Comment (vepadulano) [reply tree/tree/inc/ROOT/RFriendInfo.hxx:73]: Agreed!

File: tree/tree/inc/ROOT/RFriendInfo.hxx (hunk 8 above)
Comment (eguiraud) [reply tree/tree/inc/ROOT/RFriendInfo.hxx:80]: `RFriendInfo` now has a data member of type `vector<unique_ptr<_>>` which implicitly disables copy-construction and copy-assignment, so I implement them explicitly in order to not break backwards-compatibility. And implementing other constructors explicitly disables aggregate initialization (`RFriendInfo{x, y, z}`), so we need a dedicated constructor for that purpose.

**Review by vepadulano — APPROVED** [review]
Hooray for avoiding bogus data read :D

File: tree/treeplayer/src/TChainIndex.cxx (hunk 6 above)
Comment (eguiraud) [reply tree/treeplayer/src/TChainIndex.cxx]: done

**Comment by eguiraud** [comment]
clang-format wants me to do things that I don't want to do:

```diff
diff --git a/tree/tree/inc/TVirtualIndex.h b/tree/tree/inc/TVirtualIndex.h
index c4753fac14..7f38cb2ca0 100644
--- a/tree/tree/inc/TVirtualIndex.h
+++ b/tree/tree/inc/TVirtualIndex.h
@@ -45,7 +45,7 @@ public:
    virtual Long64_t       GetN()            const = 0;
    virtual TTree         *GetTree()         const {return fTree;}
    virtual void           UpdateFormulaLeaves(const TTree *parent) = 0;
-   virtual void           SetTree(TTree *T) = 0;
+   virtual void SetTree(TTree *T) = 0;

    ClassDefOverride(TVirtualIndex,1);  //Abstract interface for Tree Index
 };
diff --git a/tree/treeplayer/inc/TChainIndex.h b/tree/treeplayer/inc/TChainIndex.h
index 83bfba1f30..7eca7b529f 100644
--- a/tree/treeplayer/inc/TChainIndex.h
+++ b/tree/treeplayer/inc/TChainIndex.h
@@ -88,7 +88,7 @@ public:
    virtual Long64_t       GetN()            const {return fEntries.size();}
    virtual Bool_t         IsValidFor(const TTree *parent);
    virtual void           UpdateFormulaLeaves(const TTree *parent);
-   virtual void           SetTree(TTree *T);
+   virtual void SetTree(TTree *T);

    ClassDef(TChainIndex,1)  //A Tree Index with majorname and minorname.
 };
diff --git a/tree/treeplayer/inc/TTreeIndex.h b/tree/treeplayer/inc/TTreeIndex.h
index d56ad5c3ac..218f21403a 100644
--- a/tree/treeplayer/inc/TTreeIndex.h
+++ b/tree/treeplayer/inc/TTreeIndex.h
@@ -68,7 +68,7 @@ public:
    virtual Bool_t         IsValidFor(const TTree *parent);
    virtual void           Print(Option_t *option="") const;
    virtual void           UpdateFormulaLeaves(const TTree *parent);
-   virtual void           SetTree(TTree *T);
+   virtual void SetTree(TTree *T);

    ClassDef(TTreeIndex,2);  //A Tree Index with majorname and minorname.
 };
```

File: tree/tree/inc/ROOT/RFriendInfo.hxx (hunk 5 above)
Comment (pcanal) [reply tree/tree/inc/ROOT/RFriendInfo.hxx]: > For the distributed case, if we stream the RFriendInfo to the tasks, I think we'll leak the tree indexes.

You we have had something like
```
std::vector<TVirtualIndex *> fTreeIndexInfos;
bool fIsOwning = false;
```
and a read rule setting `fIsOwning` to `true` when reading back and a condition deletion of the content.

This would avoid having to duplicate the `TTreeIndex` needlessly on the main process.

**Comment by pcanal** [comment]
> clang-format wants me to do things that I don't want to do:

clang-format does not support the correct format (the one we want), so here you have to ignore clang-format (please :) )

File: tree/tree/inc/ROOT/RFriendInfo.hxx (hunk 5 above)
Comment (eguiraud) [reply tree/tree/inc/ROOT/RFriendInfo.hxx]: I don't think the extra complication (I'm thinking mostly about the read rule and its testing) is worth it: the only drawback of the current solution is that it requires one extra copy of the tree index: if we ever need that (N+1)-th of the RAM back we can revise, but for now the priority is fixing https://github.com/root-project/root/issues/12260 .

**Comment by eguiraud** [comment]
Force-pushed to fix conflicts.
